doc/compiler/optimizer/OptimizerDesignFeatures.md

```diff
@@ -0,0 +1,306 @@
+The OMR/OpenJ9 JIT compiler has both capabilities.
+
+The mid-invocation transition from interpreter to compiled code is called [Dynamic Loop Transfer (DLT)](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/control/CompilationControl.md#dynamic-loop-transfer)
+and the mid-invocation transition from compiled code to interpreter is called On Stack Replacement (OSR).
+They are valuable tools that can be essential when building JIT compilers for highly dynamic languages.
+
+Another aspect of performance is application responsiveness, which is the reason why initial compilations in
+the OMR/OpenJ9 JIT compiler are typically done at [lower optimization levels](https://github.com/eclipse/omr/blob/cf8ddbd1adcd87e388cad6be1fd0e0c085285a29/compiler/compile/CompilationTypes.hpp#L45)
+(that are cheaper to do from a compile time perspective) whereas recompilations are done at higher optimization
+levels on a more selective basis to avoid paying a high compile time penalty for every compiled method.
+
+Also, from a start-up time perspective, it is more important to stop interpreting frequently executed methods
+by executing compiled code as soon as possible, even if the code is not of peak quality (the interpreter can
+be several times slower than compiled code even at lowest optimization level).
+
+## Hot Path Information and Profile Guided Optimization
+
+The OMR/OpenJ9 JIT compiler has several optimizations that become more selective depending on the execution
+frequency of the code it is optimizing. Examples are several optimizations that cause code growth, such as
+[loop versioning](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/IntroLoopOptimizations.md#2-loop-versioner)
+and [specialization](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/IntroLoopOptimizations.md#4-loop-specializer), loop unrolling,
+[inlining](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/Inliner.md), and tail splitting,
+since care must be taken that the code growth and subsequent compile time increase in optimizations that
+run later are worth the run time benefit.
+
+Frequencies are also considered in some of the most expensive optimizations that are in the OMR/ OpenJ9 JIT
+compiler such as Global Register Allocation (GRA), wherein blocks and candidates that participate in the
+analysis can be chosen to keep compile time in check.
+
+The OMR/ OpenJ9 JIT compiler also has many [local optimizations](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/optimizer/LocalOptimizationsSummary.md)
+that run one extended basic block at a time, and as such it is trivial to constrain these optimizations to
+run mostly on more frequently executed basic blocks.
+
+Frequently executed basic blocks are identified by a [profiling framework](https://github.com/eclipse-openj9/openj9/tree/master/doc/compiler/jprofiling)
```

Comment:
I believe this refers to the generic apis to get value/block profiling information rather than JProfiling specifically since IProfiling can be a source of profiling information too. This can be seen in consumers such as https://github.com/eclipse/omr/blob/cf8ddbd1adcd87e388cad6be1fd0e0c085285a29/compiler/optimizer/LoopVersioner.cpp#L2726 where the profiling information is obtained without awareness of where it comes from.